compiler/lib/generate.ml

```diff
@@ -817,17 +841,39 @@ let parallel_renaming params args continuation queue =
 
 (****)
 
-let apply_fun_raw ctx f params =
+let apply_fun_raw ctx f params exact tailcall =
   let n = List.length params in
-  J.ECond
-    ( J.EBin (J.EqEq, J.EDot (f, "length"), int n)
-    , ecall f params J.N
-    , ecall
-        (runtime_fun ctx "caml_call_gen")
-        [ f; J.EArr (List.map params ~f:(fun x -> Some x)) ]
-        J.N )
-
-let generate_apply_fun ctx n =
+  let apply_directly = ecall f params J.N in
+  let apply =
+    (* We skip the arity check when we know that we have the right
+       number of parameters, since this test is expensive. *)
+    if exact
+    then apply_directly
+    else
+      J.ECond
+        ( J.EBin (J.EqEq, J.EDot (f, "length"), int n)
+        , apply_directly
+        , ecall
+            (runtime_fun ctx "caml_call_gen")
+            [ f; J.EArr (List.map params ~f:(fun x -> Some x)) ]
+            J.N )
+  in
+  if tailcall
```

Comment:
`if tailcall && effect ()` ?

Reply:
If we expect tailcall to only be true with effects, maybe use an assert instead.